src/Server.php

```diff
@@ -282,7 +282,9 @@ function ($error) use ($that, $conn, $request) {
             }
         );
 
-        if ($contentLength === 0) {
+        $upgradeConnection = $request->hasHeader('Connection') && $request->getHeaderLine('Connection') === 'Upgrade';
+
+        if (!$upgradeConnection && $contentLength === 0) {
             // If Body is empty or Content-Length is 0 and won't emit further data,
             // 'data' events from other streams won't be called anymore
             $stream->emit('end');
```

Comment:
I do see the issue that this causes. To support upgrades, the 'end' event could  be emitted after the response is constructed and available.

As far as implementing as a separate callback, I am not sure that that makes sense as per the spec, `Upgrade` can be ignored, or other normal HTTP things can happen (like redirect, `100-continue`, etc.).